language/move-vm/runtime/src/native_functions.rs

```diff
@@ -62,6 +81,40 @@ pub struct NativeContext<'a> {
     data_store: &'a mut dyn DataStore,
     gas_status: &'a GasStatus<'a>,
     resolver: &'a Resolver<'a>,
+    extensions: &'a mut NativeContextExtensions,
+}
+
+/// A data type to represent a heterogeneous collection of extensions which are available to
+/// native functions. A reference to this is passed into the session function execution
+/// entry points.
+#[derive(Default)]
+pub struct NativeContextExtensions {
```

Comment:
An alternative approach is to take in native functions as closures (`dyn FnMut`) as opposed to plain function pointers. The closures can access whatever outer contexts using `Rc`s. I wonder how that approach compares to this.

Reply:
I think this wont very well. For example, you can't have to closures which access the same mut state, you'd need to use RefCell for that. The current approach works rather fine, I've implemented async with it and tables are also mostly done. Also let me note that type indexed memory like here is a standard design pattern in C++/Move. Its type safe by construction, because you cant apply a downcast with the wrong type.